employee-rostering-frontend/src/ui/components/DataTable.tsx

```diff
@@ -0,0 +1,178 @@
+
+  abstract updateData(data: T): void;
+  abstract addData(data: T): void;
+  abstract removeData(data: T): void;
+
+  createNewRow() {
+    if (this.state.newRow === null) {
+      const dataStore: any = {};
+      const newRowComponents: JSX.Element[] = this.createNewDataRow(dataStore);
+      const newRow: IRow = { cells: newRowComponents.map(c => {return {title: c}}).concat([{
+        title: this.getAddButtons(dataStore)}])};
+      this.setState({...this.state, newRow: newRow});
+    }
+  }
+
+  cancelAddingRow() {
+    this.setState({...this.state, newRow: null});
+  }
+
+  getAddButtons(dataStore: C): JSX.Element {
+    return <span>
+      <Button aria-label="Save"
+        variant={ButtonVariant.link}
+        onClick={() => {
+          if (this.isValid(dataStore)) {
+            this.addData(this.extractDataFromRow({},dataStore));
+            this.cancelAddingRow();
+          }
+        }}>
+        <SaveIcon />
+      </Button>
+      <Button aria-label="Cancel"
+        variant={ButtonVariant.link}
+        onClick={this.cancelAddingRow}>
+        <CloseIcon />
+      </Button>
+    </span>;
+  }
+
+  getEditButtons(dataStore: C, data: T, editableComponents: EditableComponent[]): JSX.Element {
+    return <EditableComponent ref={(c) => {editableComponents[editableComponents.length-1] = c as EditableComponent;}}
+      viewer={<span>
+        <Button aria-label="Edit"
+          variant={ButtonVariant.link}
+          onClick={() => {
+            editableComponents.forEach(c => c.startEditing());
+          }}>
+          <EditIcon />
+        </Button>
+        <Button aria-label="Delete"
+          variant={ButtonVariant.link}
+          onClick={() => {
+            this.removeData(data);
+          }}>
+          <TrashIcon />
+        </Button>
+      </span>}
+      editor={<span>
+        <Button aria-label="Save"
+          variant={ButtonVariant.link}
+          onClick={() => {
+            if (this.isValid(dataStore)) {
+              this.updateData(this.extractDataFromRow(data,dataStore));
+              editableComponents.forEach(c => c.stopEditing());
+            }
+          }}>
+          <SaveIcon />
+        </Button>
+        <Button aria-label="Cancel"
+          variant={ButtonVariant.link}
+          onClick={() => {
+            editableComponents.forEach(c => c.stopEditing());
+          }}>
+          <CloseIcon />
+        </Button>
+      </span>}/>;
+  }
+
+  convertDataToTableRow(data: T): IRow {
+    const dataStore: any = {};
+    const viewers = this.displayDataRow(data);
+    const editors = this.editDataRow(dataStore, data);
+    const length = viewers.length
+    const editableComponents: EditableComponent[] = new Array(length + 1);
+    const cellContents = viewers.map((viewer, index) => {return { title:
+      <EditableComponent viewer={viewer}
+        editor={editors[index]}
+        ref={(c) => editableComponents[index] = c as EditableComponent}
```

Comment:
@Christopher-Chianelli Can you refactor it without using refs?
> In the typical React dataflow, props are the only way that parent components interact with their children. To modify a child, you re-render it with new props.

https://reactjs.org/docs/refs-and-the-dom.html